theano/gpuarray/opt.py

```diff
@@ -289,17 +292,17 @@ def apply(self, fgraph):
         for i in fgraph.variables:
             if isinstance(i, theano.Constant):
                 mapping[i] = i
-
-        for lopt in (self.local_optimizers_all +
-                     self.local_optimizers_map.get(type(node.op), []) +
-                     self.local_optimizers_map.get(node.op, [])):
-                process_count.setdefault(copt, 0)
-                global_process_count.setdefault(opt, 0)
-                time_opts.setdefault(opt, 0)
-                node_created.setdefault(opt, 0)
+        for node in fgraph.toposort():
+            for lopt in (self.local_optimizers_all + 
+                         self.local_optimizers_map.get(type(node.op), []) +
+                         self.local_optimizers_map.get(node.op, [])):
+                    process_count.setdefault(lopt, 0)
+                    global_process_count.setdefault(lopt, 0)
+                    time_opts.setdefault(lopt, 0)
+                    node_created.setdefault(lopt, 0)
 
         topo_t0 = time.time()
-        q = deque(graph.io_toposort(fgraph.inputs, start_from))
+        q = deque(graph.io_toposort(fgraph.inputs, fgraph.outputs))
         io_toposort_timing.append(time.time() - topo_t0)
         nb_nodes.append(len(q))
         max_nb_nodes = max(max_nb_nodes, len(q))
```

Comment:
This bloc from line 295 to 308 is useless.

Do something like:

t = time.time()
topo = fgraph.toposort()
time_topo = time.time() -t
for node in topo:  # This is the new line 310